In [35]:
# importing dependancies
import pandas as pd
from sklearn.preprocessing import LabelEncoder

In [36]:
#loading the dataset
df = pd.read_csv("STEG_BILLING_HISTORY.csv")

C:\Users\Hp\AppData\Local\Temp\ipykernel_15552\387605103.py:2: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("STEG_BILLING_HISTORY.csv")


In [37]:
# display the first 10 lines and store in a variable
client_0_bills = df.head(10)
client_0_bills

client_id invoice_date  tarif_type  counter_number counter_statue  \
0  train_Client_0   2014-03-24          11       1335667.0              0   
1  train_Client_0   2013-03-29          11       1335667.0              0   
2  train_Client_0   2015-03-23          11       1335667.0              0   
3  train_Client_0   2015-07-13          11       1335667.0              0   
4  train_Client_0   2016-11-17          11       1335667.0              0   
5  train_Client_0   2017-07-17          11       1335667.0              0   
6  train_Client_0   2018-12-07          11       1335667.0              0   
7  train_Client_0   2019-03-19          11       1335667.0              0   
8  train_Client_0   2011-07-22          11       1335667.0              0   
9  train_Client_0   2011-11-22          11       1335667.0              0   

   counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0           203               8.0                    1                    82   
1           203               6.0                    1                  1200   
2           203               8.0                    1                   123   
3           207               8.0                    1                   102   
4           207               9.0                    1                   572   
5           207               9.0                    1                   314   
6           207               9.0                    1                   541   
7           207               9.0                    1                   585   
8           203               9.0                    1                  1200   
9           203               6.0                    1                  1082   

   consommation_level_2  consommation_level_3  consommation_level_4  \
0                     0                     0                     0   
1                   184                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   
5                     0                     0                     0   
6                     0                     0                     0   
7                     0                     0                     0   
8                   186                     0                     0   
9                     0                     0                     0   

   old_index  new_index  months_number counter_type  
0      14302      14384              4         ELEC  
1      12294      13678              4         ELEC  
2      14624      14747              4         ELEC  
3      14747      14849              4         ELEC  
4      15066      15638             12         ELEC  
5      15638      15952              8         ELEC  
6      15952      16493             12         ELEC  
7      16493      17078              8         ELEC  
8       7770       9156              4         ELEC  
9       9156      10238              4         ELEC

In [38]:
# client_0_variable data type
print(type(client_0_bills))

<class 'pandas.core.frame.DataFrame'>


In [39]:
# display general dataset info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4476749 entries, 0 to 4476748
Data columns (total 16 columns):
 #   Column                Dtype  
---  ------                -----  
 0   client_id             object 
 1   invoice_date          object 
 2   tarif_type            int64  
 3   counter_number        float64
 4   counter_statue        object 
 5   counter_code          int64  
 6   reading_remarque      float64
 7   counter_coefficient   int64  
 8   consommation_level_1  int64  
 9   consommation_level_2  int64  
 10  consommation_level_3  int64  
 11  consommation_level_4  int64  
 12  old_index             int64  
 13  new_index             int64  
 14  months_number         int64  
 15  counter_type          object 
dtypes: float64(2), int64(10), object(4)
memory usage: 546.5+ MB


- The dataset has 4476749 rows and 16 columns.
- The dataset has 4 categorical features
- The dataset consumes 546.5+ MB

In [40]:
df.isnull().sum()

client_id                  0
invoice_date               0
tarif_type                 0
counter_number            48
counter_statue             0
counter_code               0
reading_remarque        4531
counter_coefficient        0
consommation_level_1       0
consommation_level_2       0
consommation_level_3       0
consommation_level_4       0
old_index                  0
new_index                  0
months_number              0
counter_type               0
dtype: int64

In [41]:
#handling missing values
# Impute missing values in counter_number with its median
df['counter_number'].fillna(df['counter_number'].median(), inplace=True)

# Impute missing values in reading_remarque with its median
df['reading_remarque'].fillna(df['reading_remarque'].median(), inplace=True)


C:\Users\Hp\AppData\Local\Temp\ipykernel_15552\4051275455.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['counter_number'].fillna(df['counter_number'].median(), inplace=True)
C:\Users\Hp\AppData\Local\Temp\ipykernel_15552\4051275455.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves a

In [42]:
df.isnull().sum()

client_id               0
invoice_date            0
tarif_type              0
counter_number          0
counter_statue          0
counter_code            0
reading_remarque        0
counter_coefficient     0
consommation_level_1    0
consommation_level_2    0
consommation_level_3    0
consommation_level_4    0
old_index               0
new_index               0
months_number           0
counter_type            0
dtype: int64

#why imputation to handle missing values.
- counter_number column seems important (likely an identifier for the meter).

- Data Retention: Imputation keeps all 4.48M rows intact, ensuring no loss of potentially valuable data.
- Low Missing Percentage: With only 0.001% missing in counter_number and 0.1% in reading_remarque, the imputed values are unlikely to skew analysis of the data significantly.

In [43]:
# run a descriptive analysis on numeric features
numeric_features = df.select_dtypes(include=['int64', 'float64']) #selecting numeric features

In [44]:
#running descriptive analysis
descriptive_stats = numeric_features.describe()
print(descriptive_stats)

         tarif_type  counter_number  counter_code  reading_remarque  \
count  4.476749e+06    4.476749e+06  4.476749e+06      4.476749e+06   
mean   2.012804e+01    1.230582e+11  1.724884e+02      7.322400e+00   
std    1.347256e+01    1.657267e+12  1.338871e+02      1.571212e+00   
min    8.000000e+00    0.000000e+00  0.000000e+00      5.000000e+00   
25%    1.100000e+01    1.211080e+05  5.000000e+00      6.000000e+00   
50%    1.100000e+01    4.945610e+05  2.030000e+02      8.000000e+00   
75%    4.000000e+01    1.115161e+06  2.070000e+02      9.000000e+00   
max    4.500000e+01    2.798115e+13  6.000000e+02      4.130000e+02   

       counter_coefficient  consommation_level_1  consommation_level_2  \
count         4.476749e+06          4.476749e+06          4.476749e+06   
mean          1.003040e+00          4.109795e+02          1.093225e+02   
std           3.083466e-01          7.573080e+02          1.220123e+03   
min           0.000000e+00          0.000000e+00          0.0000

In [45]:
# select bill records for the client with id = 'train_Client_0' using 2 methods
client_0_bills = df[df['client_id'] == 'train_Client_0']
client_0_bills

client_id invoice_date  tarif_type  counter_number counter_statue  \
0   train_Client_0   2014-03-24          11       1335667.0              0   
1   train_Client_0   2013-03-29          11       1335667.0              0   
2   train_Client_0   2015-03-23          11       1335667.0              0   
3   train_Client_0   2015-07-13          11       1335667.0              0   
4   train_Client_0   2016-11-17          11       1335667.0              0   
5   train_Client_0   2017-07-17          11       1335667.0              0   
6   train_Client_0   2018-12-07          11       1335667.0              0   
7   train_Client_0   2019-03-19          11       1335667.0              0   
8   train_Client_0   2011-07-22          11       1335667.0              0   
9   train_Client_0   2011-11-22          11       1335667.0              0   
10  train_Client_0   2008-11-25          11       1335667.0              0   
11  train_Client_0   2012-07-23          11       1335667.0              0   
12  train_Client_0   2012-11-29          11       1335667.0              0   
13  train_Client_0   2013-07-22          11       1335667.0              0   
14  train_Client_0   2014-07-18          11       1335667.0              0   
15  train_Client_0   2014-11-24          11       1335667.0              0   
16  train_Client_0   2015-11-18          11       1335667.0              0   
17  train_Client_0   2012-03-27          11       1335667.0              0   
18  train_Client_0   2011-03-30          11       1335667.0              0   
19  train_Client_0   2013-02-12          11       1335667.0              0   
20  train_Client_0   2010-07-22          11       1335667.0              0   
21  train_Client_0   2010-11-24          11       1335667.0              0   
22  train_Client_0   2005-10-17          11       1335667.0              0   
23  train_Client_0   2006-02-24          11       1335667.0              0   
24  train_Client_0   2006-06-23          11       1335667.0              0   
25  train_Client_0   2006-10-18          11       1335667.0              0   
26  train_Client_0   2007-06-27          11       1335667.0              0   
27  train_Client_0   2007-10-25          11       1335667.0              0   
28  train_Client_0   2007-02-26          11       1335667.0              0   
29  train_Client_0   2008-07-28          11       1335667.0              0   
30  train_Client_0   2009-01-04          11       1335667.0              0   
31  train_Client_0   2009-07-27          11       1335667.0              0   
32  train_Client_0   2009-11-24          11       1335667.0              0   
33  train_Client_0   2010-03-29          11       1335667.0              0   
34  train_Client_0   2008-01-04          11       1335667.0              0   

    counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0            203               8.0                    1                    82   
1            203               6.0                    1                  1200   
2            203               8.0                    1                   123   
3            207               8.0                    1                   102   
4            207               9.0                    1                   572   
5            207               9.0                    1                   314   
6            207               9.0                    1                   541   
7            207               9.0                    1                   585   
8            203               9.0                    1                  1200   
9            203               6.0                    1                  1082   
10           203               6.0                    1                   174   
11           203               6.0                    1                   628   
12           203               6.0                    1                   886   
13           203               8.0                    1           

In [46]:
client0_bills = df.query("client_id == 'train_Client_0'")
client0_bills

client_id invoice_date  tarif_type  counter_number counter_statue  \
0   train_Client_0   2014-03-24          11       1335667.0              0   
1   train_Client_0   2013-03-29          11       1335667.0              0   
2   train_Client_0   2015-03-23          11       1335667.0              0   
3   train_Client_0   2015-07-13          11       1335667.0              0   
4   train_Client_0   2016-11-17          11       1335667.0              0   
5   train_Client_0   2017-07-17          11       1335667.0              0   
6   train_Client_0   2018-12-07          11       1335667.0              0   
7   train_Client_0   2019-03-19          11       1335667.0              0   
8   train_Client_0   2011-07-22          11       1335667.0              0   
9   train_Client_0   2011-11-22          11       1335667.0              0   
10  train_Client_0   2008-11-25          11       1335667.0              0   
11  train_Client_0   2012-07-23          11       1335667.0              0   
12  train_Client_0   2012-11-29          11       1335667.0              0   
13  train_Client_0   2013-07-22          11       1335667.0              0   
14  train_Client_0   2014-07-18          11       1335667.0              0   
15  train_Client_0   2014-11-24          11       1335667.0              0   
16  train_Client_0   2015-11-18          11       1335667.0              0   
17  train_Client_0   2012-03-27          11       1335667.0              0   
18  train_Client_0   2011-03-30          11       1335667.0              0   
19  train_Client_0   2013-02-12          11       1335667.0              0   
20  train_Client_0   2010-07-22          11       1335667.0              0   
21  train_Client_0   2010-11-24          11       1335667.0              0   
22  train_Client_0   2005-10-17          11       1335667.0              0   
23  train_Client_0   2006-02-24          11       1335667.0              0   
24  train_Client_0   2006-06-23          11       1335667.0              0   
25  train_Client_0   2006-10-18          11       1335667.0              0   
26  train_Client_0   2007-06-27          11       1335667.0              0   
27  train_Client_0   2007-10-25          11       1335667.0              0   
28  train_Client_0   2007-02-26          11       1335667.0              0   
29  train_Client_0   2008-07-28          11       1335667.0              0   
30  train_Client_0   2009-01-04          11       1335667.0              0   
31  train_Client_0   2009-07-27          11       1335667.0              0   
32  train_Client_0   2009-11-24          11       1335667.0              0   
33  train_Client_0   2010-03-29          11       1335667.0              0   
34  train_Client_0   2008-01-04          11       1335667.0              0   

    counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0            203               8.0                    1                    82   
1            203               6.0                    1                  1200   
2            203               8.0                    1                   123   
3            207               8.0                    1                   102   
4            207               9.0                    1                   572   
5            207               9.0                    1                   314   
6            207               9.0                    1                   541   
7            207               9.0                    1                   585   
8            203               9.0                    1                  1200   
9            203               6.0                    1                  1082   
10           203               6.0                    1                   174   
11           203               6.0                    1                   628   
12           203               6.0                    1                   886   
13           203               8.0                    1           

In [47]:
#transform 'counter_type' feature to a numeric variable 
le = LabelEncoder()

df['counter_type_numeric'] = le.fit_transform(df['counter_type'])
df

client_id invoice_date  tarif_type  counter_number  \
0            train_Client_0   2014-03-24          11       1335667.0   
1            train_Client_0   2013-03-29          11       1335667.0   
2            train_Client_0   2015-03-23          11       1335667.0   
3            train_Client_0   2015-07-13          11       1335667.0   
4            train_Client_0   2016-11-17          11       1335667.0   
...                     ...          ...         ...             ...   
4476744  train_Client_99998   2005-08-19          10       1253571.0   
4476745  train_Client_99998   2005-12-19          10       1253571.0   
4476746  train_Client_99999   1996-09-25          11        560948.0   
4476747  train_Client_99999   1996-05-28          11        560948.0   
4476748  train_Client_99999   1996-01-25          11        560948.0   

        counter_statue  counter_code  reading_remarque  counter_coefficient  \
0                    0           203               8.0                    1   
1                    0           203               6.0                    1   
2                    0           203               8.0                    1   
3                    0           207               8.0                    1   
4                    0           207               9.0                    1   
...                ...           ...               ...                  ...   
4476744              0           202               9.0                    1   
4476745              0           202               6.0                    1   
4476746              0           203               6.0                    1   
4476747              0           203               6.0                    1   
4476748              0           203               6.0                    1   

         consommation_level_1  consommation_level_2  consommation_level_3  \
0                          82                     0                     0   
1                        1200                   184                     0   
2                         123                     0                     0   
3                         102                     0                     0   
4                         572                     0                     0   
...                       ...                   ...                   ...   
4476744                   400                   135                     0   
4476745                   200                     6                     0   
4476746                   259                     0                     0   
4476747                   603                     0                     0   
4476748                   516                     0                     0   

         consommation_level_4  old_index  new_index  months_number  \
0                           0      14302      14384              4   
1                           0      12294      13678              4   
2                           0      14624      14747              4   
3                           0      14747      14849              4   
4                           0      15066      15638             12   
...                       ...        ...        ...            ...   
4476744                     0       3197       3732              8   
4476745                     0       3732       3938              4   
4476746                     0      13884      14143              4   
4476747                     0      13281      13884              4   
4476748                     0      12765      13281              4   

        counter_type  counter_type_numeric  
0               ELEC                     0  
1               ELEC                     0  
2               ELEC                     0  
3               ELEC                     0  
4               ELEC                     0  
...              ...                   ...  
4476744         ELEC                     0  
4476745         ELEC                     0  
4476746         ELEC             

In [49]:
# delete 'counter_statue' feature

df.drop(columns=['counter_statue'], inplace=True)
df.head()

client_id invoice_date  tarif_type  counter_number  counter_code  \
0  train_Client_0   2014-03-24          11       1335667.0           203   
1  train_Client_0   2013-03-29          11       1335667.0           203   
2  train_Client_0   2015-03-23          11       1335667.0           203   
3  train_Client_0   2015-07-13          11       1335667.0           207   
4  train_Client_0   2016-11-17          11       1335667.0           207   

   reading_remarque  counter_coefficient  consommation_level_1  \
0               8.0                    1                    82   
1               6.0                    1                  1200   
2               8.0                    1                   123   
3               8.0                    1                   102   
4               9.0                    1                   572   

   consommation_level_2  consommation_level_3  consommation_level_4  \
0                     0                     0                     0   
1                   184                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   

   old_index  new_index  months_number counter_type  counter_type_numeric  
0      14302      14384              4         ELEC                     0  
1      12294      13678              4         ELEC                     0  
2      14624      14747              4         ELEC                     0  
3      14747      14849              4         ELEC                     0  
4      15066      15638             12         ELEC                     0